lib/model/rwfolder.go

```diff
@@ -524,9 +533,26 @@ nextFile:
 
 		// Verify that the thing we are handling lives inside a directory,
 		// and not a symlink or empty space.
-		if err := osutil.TraversesSymlink(f.fs, filepath.Dir(fi.Name)); err != nil {
-			f.newError("traverses q", fi.Name, err)
-			continue
+		if tErr := osutil.TraversesSymlink(f.fs, filepath.Dir(fi.Name)); tErr != nil {
+			if !fs.IsNotExist(tErr.Err) {
+				f.newError("traverses q", fi.Name, tErr)
+				continue
+			}
+			// issues #114 and #4475: This works around a race condition
+			// between two devices, when one device removes a directory and the
+			// other creates a file in it. However that happens, we end up with
+			// a directory for "foo" with the delete bit, but a file "foo/bar"
+			// that we want to sync. We never create the directory, and hence
+			// fail to create the file and end up looping forever on it. This
+			// breaks that by creating the directory and scheduling a scan,
+			// where it will be found and the delete bit on it removed.  The
+			// user can then clean up as they like...
+			l.Debugln("%v resurrecting parent directory of %v: %v", f, fi.Name, tErr.Path)
+			if err := f.fs.MkdirAll(filepath.Dir(fi.Name), 0755); err != nil {
+				f.newError("resurrecting parent dir", fi.Name, err)
+				continue
+			}
+			scanChan <- tErr.Path
```

Comment:
Why this whole lot is under TraversesSymlink if?

Reply:
Because if it returns a nil error there is nothing to do, if it reports a missing parent directory, we recreate and scan it.

Reply:
It's a bit strange to rely on traverses symlink implementation to validate that the whole path is there. What if it checks if there is no symlinks up to the point in the path where something actually exists?

Reply:
I see that the naming is a bit confusing, I don't understand the question though: That is exactly what it does: If it encounters a symlink up to the missing directory, it returns an error signifying there is a symlink -> abort. If it does not, it returns an error signifying that a parent is missing (before it would just return nil), and we create the directory.
It would probably better be named something like "PathTraverseCheck" as already before, it also reports an error when the path contains a file instead of a directory. Separating these checks into their own function would mean iterating several times over the path components, which while cheap still seems bad.

Reply:
My point is, I could easily see this function unintentionally changing the way it behaves and only checking if no symlinks exist up to the point the final path exists.

If you check for /foo/bar/baz and only /foo exists as a non-link, I wouldn't be surprised if this function didn't fail, as no symlinks are traversed, regardless that the final path does not exist.

Lstating the final directory you land in is not hard, given, you're mkdirilall'ing anyway.
Alternative would be to change name of TraversesSymlinks to explain that it also checks existance of the destination.

Reply:
Ok, I am leaving TraversesSymlink alone.